### PR TITLE
feat: support images via OpenAI Vision model: gpt-4-vision-preview

### PR DESCRIPTION
**Pull request by cycai0312**
Send image to GPT and let it analyze it

### CODE REVIEW

File: model/gpt4vision.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+	"encoding/base64"
+	"io"
+	"os"
+	"regexp"
+	"strings"
+
+	"github.com/sashabaranov/go-openai"
+)
+
+func encodeImage(imagePath string) (string, error) {
```
Comment (hsluoyz) [inline model/gpt4vision.go]: `encodeImage` is not used anywhere?

File: model/gpt4vision.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		var role string
+		if message.Author == "AI" {
+			role = openai.ChatMessageRoleAssistant
+		} else if message.Author == "System" {
+			role = openai.ChatMessageRoleSystem
+		} else {
+			role = openai.ChatMessageRoleUser
+		}
+
+		message.Text = strings.Replace(message.Text, "&nbsp;", " ", -1)
```
Comment (hsluoyz) [inline model/gpt4vision.go]: Why remove all nbsp and `<br>` ?

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: model/gpt4vision.go (hunk 1 above)
Comment (cycai0312) [reply model/gpt4vision.go]: I have deleted the code block

File: model/gpt4vision.go (hunk 2 above)
Comment (cycai0312) [reply model/gpt4vision.go]: There are some HTML strings in the message when a user manually types the text. I have to delete the extra strings.

File: model/local.go [hunk 3]
```diff
@@ -108,6 +114,7 @@ func (p *LocalModelProvider) QueryText(question string, writer io.Writer, histor
 				TopP:             topP,
 				FrequencyPenalty: frequencyPenalty,
 				PresencePenalty:  presencePenalty,
+				MaxTokens:        4096,
```
Comment (hsluoyz) [inline model/local.go]: Don't impact other providers

File: model/gpt4vision.go (hunk 2 above)
Comment (hsluoyz) [reply model/gpt4vision.go]: Extract out a function then

File: go.sum [hunk 4]
```diff
@@ -585,6 +585,8 @@ github.com/ryanuber/columnize v0.0.0-20160712163229-9b3edd62028f/go.mod h1:sm1tb
 github.com/samuel/go-zookeeper v0.0.0-20190923202752-2cc03de413da/go.mod h1:gi+0XIa01GRL2eRQVjQkKGqKF3SF9vZR/HnPullcV2E=
 github.com/sashabaranov/go-openai v1.12.0 h1:aRNHH0gtVfrpIaEolD0sWrLLRnYQNK4cH/bIAHwL8Rk=
 github.com/sashabaranov/go-openai v1.12.0/go.mod h1:lj5b/K+zjTSFxVLijLSTDZuP7adOgerWeFyZLUhAKRg=
+github.com/sashabaranov/go-openai v1.17.10 h1:ybvWN+d/rgEK/64U6dsjnOQ9AUya2wBoJKj3Wuaonqo=
```
Comment (hsluoyz) [inline go.sum:586]: run `go mod tidy`

**Review by hsluoyz — CHANGES_REQUESTED** [review]

**Comment by hsluoyz** [comment]
Replaced by: https://github.com/casibase/casibase/pull/701

**Pull request closed without merging**